util/fibers/synchronization.h

```diff
@@ -405,15 +408,17 @@ class BlockingCounter {
     }
 
     // I suspect all memory order accesses here could be "relaxed" but I do not bother.
-    void Wait() {
-      ec_.await([this] {
-        auto cnt = count_.load(std::memory_order_acquire);
-        return cnt == 0 || (cnt & (1ULL << 63));
+    bool Wait() {
+      uint64_t cnt;
+      ec_.await([this, &cnt] {
+        cnt = count_.load(std::memory_order_acquire);
+        return cnt == 0 || (cnt & kCancelFlag);
       });
+      return (cnt & kCancelFlag) == 0;
     }
 
     void Cancel() {
-      count_.fetch_or(1ULL << 63, std::memory_order_acquire);
+      count_.fetch_or(kCancelFlag, std::memory_order_acq_rel);
       ec_.notifyAll();
     }
 
```

Comment:
`Dec` below is acquire-release, why is this then acquire only? (we actually need release here, see comment below)